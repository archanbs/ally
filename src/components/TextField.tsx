import React from "react";

export type ITextFieldProps = {
  label: string;
  id: string;
  placeholder?: string;
  value?: string;
  required?: boolean;
  suffix?: string;
  type?: "text" | "number";
  width?: number;
  description?: string;
  onChange: (val: string) => void;
  min?: number;
  max?: number;
  step?: number;
  errorMessage?: string;
};

export const TextField = ({
  label,
  id,
  placeholder,
  value,
  required,
  type,
  suffix,
  width,
  description,
  onChange,
  min,
  max,
  step,
  errorMessage,
}: ITextFieldProps) => (
  <div className="mb-4">
    <label className="font-semibold" htmlFor={id}>
      {label}
    </label>
    <br />
    <input
      id={id}
      name={id}
      className="border border-gray-400 p-1 px-2 rounded mr-1"
      style={{ width: `${width}px` }}
      type={type}
      placeholder={placeholder}
      required={required}
      value={value}
      onChange={(ev) => {
        ev.preventDefault();
        onChange(ev.target.innerHTML);
      }}
      min={min}
      max={max}
      step={step}
    />
    <span>{suffix}</span>
    <br />
    {description && <div className="text-[12px] text-[#4f4f4f]">{description}</div>}
    {errorMessage && <span className="text-[12px] text-[red]">{errorMessage}</span>}
  </div>
);
