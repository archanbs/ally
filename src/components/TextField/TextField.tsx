import React from "react";

export type ITextFieldProps = {
  label: string;
  id?: string;
  placeholder?: string;
  value?: string;
  required?: boolean;
  suffix?: string;
  type?: "text" | "number";
  width?: number;
  description?: string;
  onChange: (val: string) => void;
  min?: number;
  max?: number;
  step?: number;
  errorMessage?: string;
};

export const TextField = ({
  label,
  id,
  placeholder,
  value,
  required,
  type,
  suffix,
  width,
  description,
  onChange,
  min,
  max,
  step,
  errorMessage,
}: ITextFieldProps) => {
  const defaultId = id ?? label.split(" ").join("-");

  return (
    <div className="mb-4">
      <label className="font-semibold" htmlFor={defaultId}>
        {label}
      </label>
      <br />
      <input
        id={defaultId}
        name={defaultId}
        className="border border-gray-400 p-1 px-2 rounded mr-1"
        style={{ width: `${width}px` }}
        type={type}
        placeholder={placeholder}
        required={required}
        value={value}
        onChange={(ev) => {
          ev.preventDefault();
          onChange(ev.target.value);
        }}
        min={min}
        max={max}
        step={step}
      />
      <span>{suffix}</span>
      <br />
      {description && <div className="text-[12px] text-[#4f4f4f]">{description}</div>}
      {errorMessage && <span className="text-[12px] text-[red]">{errorMessage}</span>}
    </div>
  );
};
