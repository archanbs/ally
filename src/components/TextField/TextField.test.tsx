import React from "react";
import renderer from "react-test-renderer";
import { fireEvent, render, screen } from "@testing-library/react";
import { TextField } from "./TextField";

test("Text field is rendered correctly", () => {
  const compTree = renderer
    .create(
      <TextField
        label="textfield"
        placeholder="Enter input"
        description="Enter input"
        required
        id="textfield-1"
        onChange={() => console.log("onchange")}
      />
    )
    .toJSON();

  expect(compTree).toMatchSnapshot();
});

test("Text field onchange triggers the callback function", async () => {
  const obj = {
    changeCallback: function () {
      console.log("change callback called");
    },
  };
  jest.spyOn(obj, "changeCallback");

  render(
    <TextField label="Text field" placeholder="Enter input" description="Enter input" required id="textfield-1" onChange={obj.changeCallback} />
  );
  const textField = screen.getByLabelText("Text field");
  expect(textField).toBeInTheDocument();

  fireEvent.change(textField, { target: { value: "text value" } });
  expect(obj.changeCallback).toHaveBeenCalled();
});
