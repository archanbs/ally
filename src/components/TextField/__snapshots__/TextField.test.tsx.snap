// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Text field is rendered correctly 1`] = `
<div
  className="mb-4"
>
  <label
    className="font-semibold"
    htmlFor="textfield-1"
  >
    textfield
  </label>
  <br />
  <input
    className="border border-gray-400 p-1 px-2 rounded mr-1"
    id="textfield-1"
    name="textfield-1"
    onChange={[Function]}
    placeholder="Enter input"
    required={true}
    style={
      Object {
        "width": "undefinedpx",
      }
    }
  />
  <span />
  <br />
  <div
    className="text-[12px] text-[#4f4f4f]"
  >
    Enter input
  </div>
</div>
`;
